src/core/thread/anycast_locator.cpp

```diff
@@ -0,0 +1,163 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements Anycast Locator functionality.
+ */
+
+#include "anycast_locator.hpp"
+
+#if OPENTHREAD_CONFIG_TMF_ANYCAST_LOCATOR_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/instance.hpp"
+#include "common/locator_getters.hpp"
+#include "thread/thread_tlvs.hpp"
+#include "thread/uri_paths.hpp"
+
+namespace ot {
+
+AnycastLocator::AnycastLocator(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mCallback(nullptr)
+    , mContext(nullptr)
+    , mAnycastLocate(UriPath::kAnycastLocate, HandleAnycastLocate, this)
+
+{
+    Get<Tmf::Agent>().AddResource(mAnycastLocate);
+}
+
+Error AnycastLocator::Locate(const Ip6::Address &aAnycastAddress, Callback aCallback, void *aContext)
+{
+    Error            error   = kErrorNone;
+    Coap::Message *  message = nullptr;
+    Ip6::MessageInfo messageInfo;
+
+    VerifyOrExit((aCallback != nullptr) && Get<Mle::Mle>().IsAnycastLocator(aAnycastAddress),
+                 error = kErrorInvalidArgs);
+
+    if (mCallback != nullptr)
+    {
+        // We set the `mCallback` to `nullptr` first and then abort the
+        // ongoing CoAP transaction (if any). Note that the abort method
+        // call can invoke the response callback `HandleResponse()` from
+        // which we invoke the `mCallback`.
+
+        mCallback = nullptr;
```

Comment:
```suggestion
```
I think we should call `mCallback` on abort. 

More over, is it better to return `kErrorBusy` in such case (allow only one outstanding query)?

Reply:
I intentionally went for this model over returning  `kErrorBusy` to keep the implementation simple (and small for MTD).

First implementation I had it return `kErrorBusy`, and it adds a lot of additional code and overhead, we need a new `Cancel()` method and then a related public API for this and then a CLI command to cancel an ongoing `locate` command (note that if we try to locate  an anycast address that is not present within mesh,  the existing CoAP tx parameters/mechansim (which I opted to go for to keep it simple) will try hard to retx the message (CoAP has long random wait time and up to 4 retx attempts before giving up which can take up to a minute or more).

I am actually considering to remove the public OT APIs for this (and its CLI support) and keep it as internal feature only.

Reply:
I think we need to guarantee that if `otThreadLocateAnycastDestination` returned `OT_ERROR_NONE`, then `mCallback` **MUST** be called in all cases, including aborted.

Thoughts?

Reply:
For most cases I agree. But for general case, I would say it depends on the method and how different methods are defined and intended to be used.

For example, if a set of  methods/APIs are intended for use by single module and there is a `Cancel()` method to cancel a previous request (which had its own async callback), then it can be fine for the callback not be invoked on a explicit `Cancel()` call (it's basically like notifying the same module that it called the `Cancel()`). 

BTW, I did not mention it in comment above, but I was/am going to change it in the next push to report error in callback.